### PR TITLE
[DataPipe] Improve Mapper to accept input/output index when apply fn

### PR DESCRIPTION
**Pull request by ejguan**
Fixes https://github.com/facebookexternal/torchdata/issues/135

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#64697 [DataPipe] Improve Mapper to accept input/output index when apply fn**

Differential Revision: [D30841472](https://our.internmc.facebook.com/intern/diff/D30841472)

### CODE REVIEW

**Comment by ejguan** [comment]
## Summary of proposed API
```py
dp.map(fn, input_col, output_col)
```

For the case of `list` or `tuple` as input
|                                                                                  | Single/Multiple Input Column (`input_col=0`/`input_col=[1, 0]`)                                           |
|----------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------|
| Default output column (`output_col=None`)                                        | The original/left-most column (0/1) is replaced                                                           |
| Single specified output column (`output_col=2`)                                  | The specified column (2) is replaced                                                                      |
| New output column (`output_col=-1`)                                              | One new column is appended                                                                                |
| Multiple output and multiple specified output column (`output_col=[0, 1]`)       | The specified columns (0 and 1) are replaced by expanded result (size should match)                       |
| Multiple output and new output columns (`output_col=[-1, -1]`)                   | Two new columns are appended by expanded result (size should match)                                       |
| Multiple output and mixing with replacement and appending (`output_col=[-1, 1]`) | The first element in result is appended and the second element in result replaces the original column (1) |

For the case of dict as input, the difference is using key rather than index to specify the `input_col` and `output_col`.

File: torch/utils/data/datapipes/iter/callable.py [hunk 1]
```diff
@@ -44,43 +50,102 @@ class MapperIterDataPipe(IterDataPipe[T_co]):
...
+        if self.input_col is None and self.output_col is None:
+            return res
+
+        if isinstance(data, tuple):
+            t_flag = True
+            data = list(data)
+        else:
+            t_flag = False
+            # Deepcopy data to prevent the original date modified. E.g. list, dict
+            data = copy.deepcopy(data)
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/callable.py]: I am not sure this is the most elegant way to protect original data

File: torch/utils/data/datapipes/iter/callable.py [hunk 2]
```diff
@@ -44,43 +50,104 @@ class MapperIterDataPipe(IterDataPipe[T_co]):
...
+            return res
+
+        # Copy tuple to list and run in-place modification because tuple is immutable.
+        if isinstance(data, tuple):
+            t_flag = True
+            data = list(data)
+        else:
+            t_flag = False
+            # Deepcopy data to prevent the original data modified. E.g. list, dict
+            data = copy.deepcopy(data)
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/callable.py:100]: I am not sure this is the most elegant way to protect original data

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30841472).

**Comment by ejguan** [comment]
|                                                        | map without index                                              | map with index                                                |
|--------------------------------------------------------|----------------------------------------------------------------|---------------------------------------------------------------|
| Replace the original column                            | `dp.map(lambda d: (-d[0], *d[1:]))`                            | `dp.map(lambda d: -d, 0)`                                     |
| Replace the left-most column (1) with multiple outputs | `dp.map(lambda d: (d[0], (d[1] - d[0], d[1] + d[0]), *d[2:]))` | `dp.map(lambda d0, d1: (d0 - d1, d0 + d1), [1, 0])`           |
| Replace a different column                             | `dp.map(lambda d: (d[0], -d[0], *d[2:]))`                      | `dp.map(lambda d: -d, 0, 1)`                                  |
| Replace multiple columns                               | `dp.map(lambda d: (d[0], d[0] - d[1], d[0] + d[1], *d[3:]))`   | `dp.map(lambda d0, d1: (d0 - d1, d0 + d1), [0, 1], [1, 2])`   |
| Append a new column                                    | `dp.map(lambda d: (*d, -d[0]))`                                | `dp.map(lambda d: -d, 0, -1)`                                 |
| Append multiple columns                                | `dp.map(lambda d: (*d, d[0] - d[1], d[0] + d[1]))`             | `dp.map(lambda d0, d1: (d0 - d1, d0 + d1), [0, 1], [-1, -1])` |
| Replace and append columns                             | `dp.map(lambda d: (d[0] + d[1], *d[1:], d[0] - d[1]))`         | `dp.map(lambda d0, d1: (d0 - d1, d0 + d1), [0, 1], [-1, 0])`  |

**Comment by ejguan** [comment]
Discussed with @wenleix about multi-return function. Currently, TA only supports to embedded multi-element result as a DataFrame into a column. (Nested DataFrame, `df["c"] = df.map(fn, columns=["a", "b"], dtype=...)`)

Even though both of us think it would be potentially useful for users to expand the multi-element result and assign them to different columns, we decide not to enable this feature for now, for the sake of BC. As it's not hard to implement, we can easily implement it in the future after we gather more feedback from community about the API design.

**Comment by ejguan** [comment]
Each column represents the value of `input_col`, each row represents the value of `output_col`. We will restrict `output_col` as a single index.

|      | None (fn=lambda d: -d)          | None (fn=lambda d: -d[0] - d[1])        | 0 (fn=lambda d: -d)  | [1, 3] (fn=lambda d0, d1: - d0 - d1)               |
|------|---------------------------------|-----------------------------------------|----------------------|----------------------------------------------------|
| None | 1 -> -1                         | (1, 2) -> -3                            | (1, 2) -> (-1, 2)    | (1, 2, 3, 4) -> ??? (1, -6, 3) or (1, 3, -6) or -6 |
| 0    | Not applicable                  | (1, 2) -> (-3, 1) or Not applicable?    | (1, 2) -> (-1, 2)    | (1, 2, 3, 4) -> (-6, 2, 3, 4)                      |
| 1    | Not applicable                  | (1, 2) -> (1, -3) or Not applicable?    | (1, 2) -> (1, -1)    | (1, 2, 3, 4) -> (1, -6, 3, 4)                      |
| -1   | 1 -> (1, -1) or Not applicable? | (1, 2) -> (1, 2, -3) or Not applicable? | (1, 2) -> (1, 2, -1) | (1, 2, 3, 4) -> (1, 2, 3, 4, -6)                   |

The parts that I am unsure about are:
- When `input_col` is `None` and input data is a tuple or list, do we accept to specify `output_col`? Personally, I think we can prevent it.
- When `input_col` is a list of indices and `output_col` is `None`, we have three options:
  - Replace the left-most index column with the result (This is my preference)
  - Replace the whole input with the result
  - Appending the result (same behavior when `output_col` is -1)

WDYT @VitalyFedyunin

File: torch/utils/data/datapipes/iter/callable.py [hunk 3]
```diff
@@ -44,43 +50,104 @@ class MapperIterDataPipe(IterDataPipe[T_co]):
...
+    def _apply_fn(self, data):
+        if self.input_col is None:
+            res = self.fn(data, *self.args, **self.kwargs)
+        elif isinstance(self.input_col, (list, tuple)):
+            args = tuple(data[col] for col in self.input_col)
+            res = self.fn(*args, *self.args, **self.kwargs)
+        else:
+            res = self.fn(data[self.input_col], *self.args, **self.kwargs)
+
+        if self.input_col is None and self.output_col is None:
```
Comment (VitalyFedyunin) [inline torch/utils/data/datapipes/iter/callable.py:90]: super nit: should be first line in the function

File: torch/utils/data/datapipes/iter/callable.py [hunk 4]
```diff
@@ -44,43 +50,104 @@ class MapperIterDataPipe(IterDataPipe[T_co]):
...
+            # Deepcopy data to prevent the original data modified. E.g. list, dict
+            data = copy.deepcopy(data)
+
+        if self.output_col is None:
+            if isinstance(self.input_col, (list, tuple)):
+                data[self.input_col[0]] = res
+            else:
+                data[self.input_col] = res
+        elif isinstance(self.output_col, (list, tuple)):
+            if not isinstance(res, Sized) or len(res) != len(self.output_col):
```
Comment (VitalyFedyunin) [inline torch/utils/data/datapipes/iter/callable.py:108]: nit: if you disable multiple outputs, make sure that [0] is still supported.

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by ejguan** [comment]
Closing this PR as a duplicate one is uploaded accidentally. https://github.com/pytorch/pytorch/pull/64951

**Pull request closed without merging**